### PR TITLE
Added inv(diag) -> (1/diag) rewrite for eye mul

### PR DESCRIPTION
**Pull request by tanish1729**
## Description
Adds the rewrite for inv(diag) -> 1/diag

## Related Issue

- [ ] Related to #573 

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Comment by tanish1729** [comment]
@ricardoV94 @jessegrabowski

File: pytensor/tensor/rewriting/linalg.py [hunk 1]
```diff
@@ -377,3 +379,119 @@ def local_lift_through_linalg(
...
+
+    # Get all non Eye inputs (scalars/matrices/vectors)
+    non_eye_inputs = list(set(inputs_to_mul) - set(eye_input))
+    return eye_input, non_eye_inputs
+
+
+@register_canonicalize
+@register_stabilize
+@node_rewriter([Blockwise])
+def rewrite_inv_for_diag_and_orthonormal(fgraph, node):
```
Comment (jessegrabowski) [inline pytensor/tensor/rewriting/linalg.py:417]: imo this should be split into two rewrites

File: tests/tensor/rewriting/test_linalg.py [hunk 2]
```diff
@@ -390,3 +390,40 @@ def test_local_lift_through_linalg(constructor, f_op, f, g_op, g):
...
 
     np.testing.assert_allclose(f1(*test_vals), f2(*test_vals), atol=1e-8)
+
+
+@pytest.mark.parametrize(
+    "shape",
+    [(), (7,), (1, 7), (7, 1), (7, 7), (3, 7, 7)],
+    ids=["scalar", "vector", "row_vec", "col_vec", "matrix", "batched_input"],
+)
+def test_inv_diag_from_eye_mul(shape):
```
Comment (jessegrabowski) [inline tests/tensor/rewriting/test_linalg.py:400]: Test for orthonormal matrix?

File: pytensor/tensor/rewriting/linalg.py [hunk 3]
```diff
@@ -377,3 +379,119 @@ def local_lift_through_linalg(
...
+    list of Variable, optional
+        List of optimized variables, or None if no optimization was performed
+    """
+    # List of useful operations : Inv, Pinv, Solve
+    core_op = node.op.core_op
+    if not (
+        isinstance(core_op, inv)
+        or isinstance(core_op, pinv)
+        or isinstance(core_op, solve)
+    ):
```
Comment (jessegrabowski) [inline pytensor/tensor/rewriting/linalg.py:450]: There are other solve Ops that could represent an inversion, notable `solve_triangular`

File: pytensor/tensor/rewriting/linalg.py [hunk 4]
```diff
@@ -377,3 +379,119 @@ def local_lift_through_linalg(
...
+        return None
+
+    # Dealing with direct inverse Ops
+    if isinstance(core_op, inv) or isinstance(core_op, pinv):
```
Comment (jessegrabowski) [inline pytensor/tensor/rewriting/linalg.py:454]:
```suggestion
    if isinstance(core_op, inv | pinv):
```

File: pytensor/tensor/rewriting/linalg.py [hunk 5]
```diff
@@ -377,3 +379,119 @@ def local_lift_through_linalg(
...
+            # To make sure input is orthonormal, we have to check that its not S
+            if input_to_inv == input_to_inv.owner.outputs[1]:
+                return None
+
+            orthonormal_input = input_to_inv
+            inverse = orthonormal_input.T
+            return [inverse]
+
+    # Dealing with solve(x, eye)
+    elif isinstance(core_op, solve):
```
Comment (jessegrabowski) [inline pytensor/tensor/rewriting/linalg.py:496]: Most of the solve logic is going to be the same as the inv/pinv, except that you just have to check that B is eye, then extract the A argument. It probably won't be best to split it out this way

File: pytensor/tensor/rewriting/linalg.py [hunk 6]
```diff
@@ -377,3 +379,119 @@ def local_lift_through_linalg(
...
+            ):
+                return None
+
+            # To make sure input is orthonormal, we have to check that its not S
+            if input_to_inv == input_to_inv.owner.outputs[1]:
+                return None
+
+            orthonormal_input = input_to_inv
+            inverse = orthonormal_input.T
+            return [inverse]
```
Comment (jessegrabowski) [inline pytensor/tensor/rewriting/linalg.py:493]:
```suggestion
            return [input_to_inv.T]
```

File: pytensor/tensor/rewriting/linalg.py [hunk 7]
```diff
@@ -377,3 +379,119 @@ def local_lift_through_linalg(
...
+            input_to_inv = node.inputs[0]
+            # Check if this input is coming from SVD with compute_uv = True
+            if not (
+                isinstance(input_to_inv.owner.op, Blockwise)
+                and isinstance(input_to_inv.owner.op.core_op, SVD)
+                and input_to_inv.owner.op.core_op.compute_uv is True
+            ):
+                return None
+
+            # To make sure input is orthonormal, we have to check that its not S
```
Comment (jessegrabowski) [inline pytensor/tensor/rewriting/linalg.py:487]:
```suggestion
            # To make sure input is orthonormal, we have to check that its not S (output order is U, S, Vh, so S is index 1)
```

File: pytensor/tensor/rewriting/linalg.py [hunk 8]
```diff
@@ -377,3 +379,119 @@ def local_lift_through_linalg(
...
+        potential_mul_input = node.inputs[0]
+        eye_non_eye_inputs = _find_diag_from_eye_mul(potential_mul_input)
+        if eye_non_eye_inputs is not None:
+            eye_input, non_eye_inputs = eye_non_eye_inputs
+
+            # Dealing with only one other input
+            if len(non_eye_inputs) != 1:
+                return None
+
+            useful_eye, useful_non_eye = eye_input[0], non_eye_inputs[0]
```
Comment (jessegrabowski) [inline pytensor/tensor/rewriting/linalg.py:465]:
```suggestion
            eye, non_eye = eye_input[0], non_eye_inputs[0]
```
Note sure what makes them "useful"

File: pytensor/tensor/rewriting/linalg.py [hunk 9]
```diff
@@ -377,3 +379,119 @@ def local_lift_through_linalg(
                 return [block_diag(*inner_matrices)]
             else:
                 raise NotImplementedError  # pragma: no cover
+
+
+def _find_diag_from_eye_mul(potential_mul_input):
```
Comment (jessegrabowski) [inline pytensor/tensor/rewriting/linalg.py:384]: Is this duplicated from the other PR?

**Review by jessegrabowski — CHANGES_REQUESTED** [review]

**Comment by jessegrabowski** [comment]
Looks like you also need to rebase from main

File: pytensor/tensor/rewriting/linalg.py (hunk 1 above)
Comment (tanish1729) [reply pytensor/tensor/rewriting/linalg.py:417]: okay i'll do that

File: pytensor/tensor/rewriting/linalg.py (hunk 9 above)
Comment (tanish1729) [reply pytensor/tensor/rewriting/linalg.py:384]: yea. that one isnt merged yet so i had to add this function in this one too

**Comment by maresb** [comment]
@tanish1729, you should be able to rebase using the command `git rebase main` to avoid deleting your branch. This way everything can stay within the same PR.

**Comment by tanish1729** [comment]
hi! yep i figured that out and was doing that but a lot of things (plenty of conflicts) got confusing so i decided to start over. 
i did eventually figure out rebasing and stuff by trying it out so i'll be sure to use that from now. thanks :D

**Pull request closed without merging**